components/board/pi/impl/board.go

```diff
@@ -101,110 +120,99 @@ func NewPigpio(ctx context.Context, name resource.Name, cfg *genericlinux.Config
 		return nil, picommon.ConvertErrorCodeToMessage(int(resCode), "gpioCfgSetInternals failed with code")
 	}
 
-	// setup
-	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	if err := initializePigpio(); err != nil {
+		return nil, err
+	}
+
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
 	piInstance := &piPigpio{
 		Named:           name.AsNamed(),
-		cfg:             cfg,
 		logger:          logger,
 		isClosed:        false,
 		interruptCtx:    cancelCtx,
 		interruptCancel: cancelFunc,
 	}
 
-	instanceMu.Lock()
-
-	// if pigpio is not initialized, only then we initialize it.
-	if !pigpioInitialized {
-		resCode = C.gpioInitialise()
-		if resCode < 0 {
-			instanceMu.Unlock()
-			// failed to init, check for common causes
-			_, err := os.Stat("/sys/bus/platform/drivers/raspberrypi-firmware")
-			if err != nil {
-				return nil, errors.New("not running on a pi")
-			}
-			if os.Getuid() != 0 {
-				return nil, errors.New("not running as root, try sudo")
-			}
-			return nil, picommon.ConvertErrorCodeToMessage(int(resCode), "error")
-		}
+	if err := piInstance.performConfiguration(ctx, nil, cfg); err != nil {
+		C.gpioTerminate()
+		logger.Debug("Pi GPIO terminated due to failed init.")
+		return nil, err
 	}
+	return piInstance, nil
+}
 
-	pigpioInitialized = true
+// TODO(RSDK-RSDK-2691): implement reconfigure.
+func (pi *piPigpio) performConfiguration(
+    _ context.Context,
+    _ resource.Dependencies,
+    conf resource.Config,
+) error {
+	cfg, err := resource.NativeConfig[*genericlinux.Config](conf)
+	if err != nil {
+		return err
+	}
 
-	initGood := false
```

Comment:
do we need to find a new place to verify whether initialization was properly done? or is this handled by perform reconfiguration?

Reply:
The only places we could possibly fail initialization are now in (what will soon become) reconfiguration. In this function, we check for errors when reconfiguration returns, and do the same things that `initGood` had been checking for.